frame/fast-unstake/src/lib.rs

```diff
@@ -324,35 +330,46 @@ pub mod pallet {
 		/// found out to have no eras to check. At the end of a check cycle, even if they are fully
 		/// checked, we don't finish the process.
 		pub(crate) fn do_on_idle(remaining_weight: Weight) -> Weight {
-			let mut eras_to_check_per_block = ErasToCheckPerBlock::<T>::get();
+			// any weight that is unaccounted for
+			let mut unaccounted_weight = Weight::from_ref_time(0);
+
+			let eras_to_check_per_block = ErasToCheckPerBlock::<T>::get();
 			if eras_to_check_per_block.is_zero() {
-				return T::DbWeight::get().reads(1)
+				return T::DbWeight::get().reads(1).saturating_add(unaccounted_weight)
 			}
 
 			// NOTE: here we're assuming that the number of validators has only ever increased,
 			// meaning that the number of exposures to check is either this per era, or less.
 			let validator_count = T::Staking::desired_validator_count();
+			let (next_batch_size, reads_from_queue) = Head::<T>::get()
+				.map_or((Queue::<T>::count().min(T::BatchSize::get()), true), |head| {
+					(head.stashes.len() as u32, false)
+				});
 
 			// determine the number of eras to check. This is based on both `ErasToCheckPerBlock`
 			// and `remaining_weight` passed on to us from the runtime executive.
-			let max_weight = |v, u| {
-				<T as Config>::WeightInfo::on_idle_check(v * u)
-					.max(<T as Config>::WeightInfo::on_idle_unstake())
+			let max_weight = |v, b| {
+				// NOTE: this potentially under-counts by up to `BatchSize` reads from the queue.
```

Comment:
Not sure how severe this is, but would it be very bad to let it over-estimate?  
Currently it should be a para-only pallet, but you never know :laughing:

Reply:
Yeah the weight of this process is already like 1/4 of the full block in some networks 🙈  rather write a bit more code to do "accurate weighing" instead of over-estimating.

Note that this code is now 100% accurate, not under-counting because of: 

```rust
.saturating_add(if reads_from_queue {
	T::DbWeight::get().reads(next_batch_size.into())
} else {
	Zero::zero()
})
```